### PR TITLE
ISPN-12498 Server Management check authorizations

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12498

### CODE REVIEW

**Comment by tristantarrant** [comment]
@pruivo this has logging and cache ignore checks too

File: server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java [hunk 1]
```diff
@@ -43,14 +45,17 @@
...
    final Map<String, DefaultCacheManager> cacheManagers;
    final Map<String, BackupRequest> backupMap;
    final Map<String, CompletionStage<Void>> restoreMap;
 
+
    public BackupManagerImpl(BlockingManager blockingManager, EmbeddedCacheManager cm,
                             Map<String, DefaultCacheManager> cacheManagers, Path dataRoot) {
       this.blockingManager = blockingManager;
       this.rootDir = dataRoot.resolve(WORKING_DIR);
+      this.cacheManager = cm;
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java:58]: can't you use one of the `DefaultCacheManager` from `this.cacheManagers`?

File: server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java [hunk 2]
```diff
@@ -43,14 +45,17 @@
    final BackupReader reader;
    final Lock backupLock;
    final Lock restoreLock;
+   private final EmbeddedCacheManager cacheManager;
    final Map<String, DefaultCacheManager> cacheManagers;
    final Map<String, BackupRequest> backupMap;
    final Map<String, CompletionStage<Void>> restoreMap;
 
+
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java:53]: nitpick: remove

File: server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java [hunk 3]
```diff
@@ -67,11 +72,13 @@ public void init() throws IOException {
...
    public Set<String> getBackupNames() {
+      SecurityActions.checkPermission(cacheManager.withSubject(Security.getSubject()), AuthorizationPermission.ADMIN);
       return new HashSet<>(backupMap.keySet());
    }
 
    @Override
    public Status getBackupStatus(String name) {
+      SecurityActions.checkPermission(cacheManager.withSubject(Security.getSubject()), AuthorizationPermission.ADMIN);
       return getBackupStatus(backupMap.get(name));
    }
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java:84]: `getBackupLocation()` is missing.

File: server/core/src/main/java/org/infinispan/server/core/backup/SecurityActions.java [hunk 4]
```diff
@@ -31,4 +35,14 @@ static GlobalConfiguration getGlobalConfiguration(final EmbeddedCacheManager cac
...
    }
+
+   static GlobalComponentRegistry getGlobalComponentRegistry(final EmbeddedCacheManager cacheManager) {
+      GetGlobalComponentRegistryAction action = new GetGlobalComponentRegistryAction(cacheManager);
+      return doPrivileged(action);
+   }
+
+   static void checkPermission(EmbeddedCacheManager cacheManager, AuthorizationPermission permission) {
+      AuthorizationHelper authzHelper = getGlobalComponentRegistry(cacheManager).getComponent(AuthorizationHelper.class);
+      authzHelper.checkPermission(cacheManager.getSubject(), AuthorizationPermission.ADMIN);
```
Comment (pruivo) [inline server/core/src/main/java/org/infinispan/server/core/backup/SecurityActions.java:46]:
```suggestion
      authzHelper.checkPermission(cacheManager.getSubject(), permission);
```

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java [hunk 5]
```diff
@@ -38,8 +40,8 @@ public ClusterResource(InvocationHelper invocationHelper) {
    public Invocations getInvocations() {
       return new Invocations.Builder()
             .invocation().methods(POST).path("/v2/cluster").withAction("stop").handleWith(this::stop)
-            .invocation().methods(GET).path("/v2/cluster/backups").handleWith(this::getAllBackupNames)
-            .invocation().methods(DELETE, GET, POST).path("/v2/cluster/backups/{backupName}").handleWith(this::backup)
+            .invocation().methods(GET, HEAD).path("/v2/cluster/backups").handleWith(this::getAllBackupNames)
+            .invocation().methods(DELETE, GET, HEAD, POST).path("/v2/cluster/backups/{backupName}").handleWith(this::backup)
             .invocation().methods(GET).path("/v2/cluster/restores").handleWith(this::getAllRestoreNames)
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:45]: `getAllRestoreNames()` isn't protected.

File: server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java [hunk 6]
```diff
@@ -65,37 +65,38 @@ public Invocations getInvocations() {
...
-         return response.build();
-      });
+      }
+      return response.build();
    }
 
    private CompletionStage<RestResponse> deleteLogger(RestRequest request) {
       TaskManager taskManager = invocationHelper.getServer().getTaskManager();
       String loggerName = request.variables().get("loggerName");
       return taskManager.runTask("@@logging@remove", new TaskContext().addParameter("loggerName", loggerName))
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java:98]: isn't required the Subject be set in the `TaskContext`?

File: server/runtime/src/main/java/org/infinispan/server/SecurityActions.java [hunk 7]
```diff
@@ -110,4 +112,9 @@ static void setInitialContextFactoryBuilder(InitialContextFactoryBuilder initial
          return null;
       });
    }
+
+   static void checkPermission(EmbeddedCacheManager cacheManager, AuthorizationPermission permission) {
+      AuthorizationHelper authzHelper = getGlobalComponentRegistry(cacheManager).getComponent(AuthorizationHelper.class);
+      authzHelper.checkPermission(cacheManager.getSubject(), AuthorizationPermission.ADMIN);
```
Comment (pruivo) [inline server/runtime/src/main/java/org/infinispan/server/SecurityActions.java:118]:
```suggestion
      authzHelper.checkPermission(cacheManager.getSubject(), permission);
```

**Review by pruivo — CHANGES_REQUESTED** [review]

File: server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java (hunk 1 above)
Comment (tristantarrant) [reply server/core/src/main/java/org/infinispan/server/core/backup/BackupManagerImpl.java:58]: I have another PR to remove multiple cache managers that I'll refresh. This makes things simpler for now here.

**Comment by pruivo** [comment]
@tristantarrant I think cross-site also needs some love :/

File: server/runtime/src/main/java/org/infinispan/server/SecurityActions.java (hunk 7 above)
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/SecurityActions.java]: Oops

File: server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java (hunk 6 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/LoggingResource.java]: Yep, also added a test for this.

File: server/core/src/main/java/org/infinispan/server/core/backup/SecurityActions.java (hunk 4 above)
Comment (tristantarrant) [reply server/core/src/main/java/org/infinispan/server/core/backup/SecurityActions.java]: Oops #2

**Comment by tristantarrant** [comment]
Repushed

**Comment by tristantarrant** [comment]
Now complete with xsite stuff

File: core/src/main/java/org/infinispan/xsite/GlobalXSiteAdminOperations.java [hunk 8]
```diff
@@ -107,6 +108,7 @@ public final String cancelPushState(@Parameter(description = "The destination si
    }
 
    public final Map<String, SiteStatus> globalStatus() {
+      SecurityActions.checkPermission(cacheManager, AuthorizationPermission.ADMIN);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/GlobalXSiteAdminOperations.java:111]: stupid question: isn't `cacheManager.withSubject(Security.getSubject())` required here in this class?

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java [hunk 9]
```diff
@@ -49,16 +51,22 @@ public Invocations getInvocations() {
       List<String> servers = restRequest.parameters().get("server");
 
       if (servers != null && !servers.isEmpty()) {
-         invocationHelper.getServer().serverStop(servers);
+         Security.doAs(restRequest.getSubject(), (PrivilegedAction) () -> {
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:54]: nitpick
```suggestion
         Security.doAs(restRequest.getSubject(), (PrivilegedAction<Void>) () -> {
```

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java [hunk 10]
```diff
@@ -49,16 +51,22 @@ public Invocations getInvocations() {
...
+            invocationHelper.getServer().serverStop(servers);
+            return null;
+         });
       } else {
-         invocationHelper.getServer().clusterStop();
+         Security.doAs(restRequest.getSubject(), (PrivilegedAction) () -> {
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:59]: nitpick
```suggestion
         Security.doAs(restRequest.getSubject(), (PrivilegedAction<Void>) () -> {
```

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java [hunk 11]
```diff
@@ -49,16 +51,22 @@ public Invocations getInvocations() {
...
+            return null;
+         });
       }
       return CompletableFuture.completedFuture(new NettyRestResponse.Builder().status(NO_CONTENT).build());
    }
 
    private CompletionStage<RestResponse> getAllBackupNames(RestRequest request) {
       BackupManager backupManager = invocationHelper.getServer().getBackupManager();
-      Set<String> names = backupManager.getBackupNames();
+      Set<String> names = Security.doAs(request.getSubject(), (PrivilegedAction<Set<String>>) () -> backupManager.getBackupNames());
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:69]: nit
```suggestion
      Set<String> names = Security.doAs(request.getSubject(), (PrivilegedAction<Set<String>>) backupManager::getBackupNames);
```

File: server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java [hunk 12]
```diff
@@ -69,7 +77,7 @@ public Invocations getInvocations() {
 
    private CompletionStage<RestResponse> getAllRestoreNames(RestRequest request) {
       BackupManager backupManager = invocationHelper.getServer().getBackupManager();
-      Set<String> names = backupManager.getRestoreNames();
+      Set<String> names = Security.doAs(request.getSubject(), (PrivilegedAction<Set<String>>) () -> backupManager.getRestoreNames());
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/ClusterResource.java:80]: nit
```suggestion
      Set<String> names = Security.doAs(request.getSubject(), (PrivilegedAction<Set<String>>) backupManager::getRestoreNames);
```

File: server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java [hunk 13]
```diff
@@ -117,7 +119,8 @@ public Invocations getInvocations() {
       if (globalXSiteAdmin == null) return CompletableFuture.completedFuture(responseBuilder.status(NOT_FOUND).build());
 
       return CompletableFuture.supplyAsync(() -> {
-         Map<String, GlobalStatus> collect = globalXSiteAdmin.globalStatus().entrySet().stream().collect(Collectors.toMap(Entry::getKey, e -> {
+         Map<String, SiteStatus> globalStatus = Security.doAs(request.getSubject(), (PrivilegedAction<Map<String, SiteStatus>>) () -> globalXSiteAdmin.globalStatus());
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/XSiteResource.java:122]: nit
```suggestion
         Map<String, SiteStatus> globalStatus = Security.doAs(request.getSubject(), (PrivilegedAction<Map<String, SiteStatus>>) globalXSiteAdmin::globalStatus);
```

File: core/src/main/java/org/infinispan/xsite/GlobalXSiteAdminOperations.java (hunk 8 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/xsite/GlobalXSiteAdminOperations.java:111]: They both work. If the `cacheManager` is decorated with a Subject, it will use that, otherwise it uses `Security.getSubject()` which uses a ThreadLocal set by the Security.doAs() call.

**Comment by tristantarrant** [comment]
@pruivo I've addressed all issues, but now I need to add checks to `SearchResource`... (mass-index, clear stats, etc)

**Comment by tristantarrant** [comment]
@gustavonalle now with SearchResource authz

File: server/rest/src/test/java/org/infinispan/rest/resources/security/SimpleSecurityDomain.java [hunk 14]
```diff
@@ -1,24 +1,30 @@
...
+   public SimpleSecurityDomain(Subject... subjects) {
+      this.subjects = new HashMap<>(subjects.length);
+      for (Subject subject : subjects) {
+         this.subjects.put(subject.getPrincipals().iterator().next().getName().toLowerCase(), subject);
+      }
    }
 
    @Override
    public Subject authenticate(String username, String password) throws SecurityException {
       if (username.equals(password)) {
```
Comment (pruivo) [inline server/rest/src/test/java/org/infinispan/rest/resources/security/SimpleSecurityDomain.java:26]: wait!? is this correct? username equals to password?

**Review by pruivo — CHANGES_REQUESTED** [review]

File: server/rest/src/test/java/org/infinispan/rest/resources/security/SimpleSecurityDomain.java (hunk 14 above)
Comment (tristantarrant) [reply server/rest/src/test/java/org/infinispan/rest/resources/security/SimpleSecurityDomain.java:26]: It's a test :)

Comment (pruivo) [reply server/rest/src/test/java/org/infinispan/rest/resources/security/SimpleSecurityDomain.java:26]: I'll blame the coffee. it takes time to kick in :rofl:

File: query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java [hunk 15]
```diff
@@ -60,6 +63,7 @@ public DistributedExecutorMassIndexer(AdvancedCache<?, ?> cache, KeyTransformati
       this.blockingManager = cache.getCacheManager().getGlobalComponentRegistry()
             .getComponent(BlockingManager.class);
       this.lock = MassIndexerLockFactory.buildLock(cache);
+      this.authorizationHelper = cache.getComponentRegistry().getComponent(AuthorizationHelper.class);
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java:66]: I'll protect it anyway

File: server/rest/src/main/java/org/infinispan/rest/resources/SearchAdminResource.java [hunk 16]
```diff
@@ -108,7 +109,7 @@ public Invocations getInvocations() {
          throw new CacheException("NotImplemented");
       } else {
          SearchStatistics searchStatistics = Search.getSearchStatistics(cache);
-         searchStatistics.getQueryStatistics().clear();
+         Security.doAs(restRequest.getSubject(), () -> searchStatistics.getQueryStatistics().clear());
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/SearchAdminResource.java:112]: why is this required? the cache already contains the subject (via `Cache.withSubject()`)

File: server/rest/src/test/java/org/infinispan/rest/resources/CacheV2ResourceTest.java [hunk 17]
```diff
@@ -643,6 +644,11 @@ public void testSearchStatistics() {
 
       // Clear all stats
       RestResponse response = join(cacheClient.clearSearchStats());
```
Comment (pruivo) [inline server/rest/src/test/java/org/infinispan/rest/resources/CacheV2ResourceTest.java:646]: shouldn't this be invoked only when `security==false`?

Comment (tristantarrant) [reply server/rest/src/test/java/org/infinispan/rest/resources/CacheV2ResourceTest.java:646]: All of the tests rely on zeroed stats so it needs to be called.

File: server/rest/src/main/java/org/infinispan/rest/resources/SearchAdminResource.java (hunk 16 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/SearchAdminResource.java:112]: Because that doesn't carry across to the SearchStatistics object

File: server/rest/src/test/java/org/infinispan/rest/resources/CacheV2ResourceTest.java (hunk 17 above)
Comment (pruivo) [reply server/rest/src/test/java/org/infinispan/rest/resources/CacheV2ResourceTest.java:646]: true, but you are invoked `clearSearchStats()` twice when `security==true`; here and other as "admin"

**Comment by tristantarrant** [comment]
Testsuite looking ok

File: query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java:66]: are you going to protect `isRunning()`?

[the remark this answers is not in the document]
Comment (tristantarrant) [reply query/src/main/java/org/infinispan/query/impl/massindex/DistributedExecutorMassIndexer.java:66]: I did that initially, but it breaks the cache info. Since it's just a read-only trait which doesn't expose any private info, it is safe as it is.

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
integrated! thanks @tristantarrant !
